// scryfall-sdk/out/Scry.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const MagicQuerier_1 = require("./util/MagicQuerier");
__export(require("./api/Cards"));
__export(require("./api/Catalog"));
__export(require("./api/Misc"));
__export(require("./api/Rulings"));
__export(require("./api/Sets"));
__export(require("./api/Symbology"));
var Cards_1 = require("./api/Cards");
exports.Cards = Cards_1.default;
var Catalog_1 = require("./api/Catalog");
exports.Catalog = Catalog_1.default;
var Misc_1 = require("./api/Misc");
exports.Misc = Misc_1.default;
var Rulings_1 = require("./api/Rulings");
exports.Rulings = Rulings_1.default;
var Sets_1 = require("./api/Sets");
exports.Sets = Sets_1.default;
var Symbology_1 = require("./api/Symbology");
exports.Symbology = Symbology_1.default;
function error() {
    return MagicQuerier_1.default.lastError;
}
exports.error = error;
function setRetry(attempts, timeout, canRetry) {
    MagicQuerier_1.default.retry = { attempts, timeout, canRetry };
}
exports.setRetry = setRetry;
